sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -634,57 +619,94 @@ pi_result piPlatformsGet(pi_uint32 NumEntries, pi_platform *Platforms,
     assert(ZeDriverCount == 1);
     ZE_CALL(zeDriverGet(&ZeDriverCount, &ZeDriver));
 
-    std::lock_guard<std::mutex> Lock(PiPlatformsCacheMutex);
-    for (const pi_platform CachedPlatform : PiPlatformsCache) {
-      if (CachedPlatform->ZeDriver == ZeDriver) {
-        Platforms[0] = CachedPlatform;
-        // if the caller sent a valid NumPlatforms pointer, set it here
-        if (NumPlatforms)
-          *NumPlatforms = 1;
+    static pi_result Res = getOrCreatePlatform(ZeDriver, Platforms);
+    if (Res != PI_SUCCESS) {
+      return Res;
+    }
+  }
+
+  if (NumPlatforms)
+    *NumPlatforms = 1;
+
+  return PI_SUCCESS;
+}
+
+static pi_result getOrCreatePlatform(ze_driver_handle_t ZeDriver,
+                                     pi_platform *Platform) {
+
+  // We will retrieve the Max CommandList Cache in this lamda function so that
+  // it only has to be executed once
+  static pi_uint32 CommandListCacheSizeValue = ([]() {
+    const char *CommandListCacheSize =
+        std::getenv("SYCL_PI_LEVEL0_MAX_COMMAND_LIST_CACHE");
+    pi_uint32 CommandListCacheSizeValue;
+    try {
+      CommandListCacheSizeValue =
+          CommandListCacheSize ? std::atoi(CommandListCacheSize) : 20000;
+    } catch (std::exception const &) {
+      zePrint("SYCL_PI_LEVEL0_MAX_COMMAND_LIST_CACHE: invalid value provided, "
+              "default set.\n");
+      CommandListCacheSizeValue = 20000;
+    }
+    return CommandListCacheSizeValue;
+  })();
 
+  try {
+    // Cache pi_platforms for reuse in the future
+    // It solves two problems;
+    // 1. sycl::device equality issue; we always return the same pi_device.
+    // 2. performance; we can save time by immediately return from cache.
+    //
+    // Note: The memory for "PiPlatformsCache" and "PiPlatformsCacheMutex" is
+    // intentionally leaked because the application may call into the SYCL
+    // runtime from a global destructor, and such a call could eventually
+    // access these variables. Therefore, there is no safe time when
+    // "PiPlatformsCache" and "PiPlatformsCacheMutex" could be deleted.
+    static auto PiPlatformsCache = new std::vector<pi_platform>;
+    static auto PiPlatformsCacheMutex = new std::mutex;
```

Comment:
No clear why you want to use `new` here. Do you have some trouble with:
```suggestion
    static std::vector<pi_platform> PiPlatformsCache;
    static std::mutex PiPlatformsCacheMutex;
```
?

Reply:
Directly above the lines of the suggested change, I left the following comment:

```
// Note: The memory for "PiPlatformsCache" and "PiPlatformsCacheMutex" is
// intentionally leaked because the application may call into the SYCL
// runtime from a global destructor, and such a call could eventually
// access these variables. Therefore, there is no safe time when
// "PiPlatformsCache" and "PiPlatformsCacheMutex" could be deleted.
```

The code is setup in this way to avoid destroying these objects before applications are done trying to access them. An example can be found in a situation where PiPlatformsGet() is called after main(). Once main() returns, these objects may be destroyed, which would lead to accessing them after they are destroyed. The downside of this is that it leaks a bit of memory but that is preferable over the program crashing.

Adding @gmlueck who identified this potential issue.

Reply:
The problem, thinking about it again is that what you want with your `new` is to avoid calling the destructor. But I am unsure how you expect `PiPlatformsCache` content usable with everything falling apart around... This sounds pretty UB-scarry...

Reply:
> thinking about it again is that what you want with your new is to avoid calling the destructor. But I am unsure how you expect PiPlatformsCache content usable with everything falling apart around.

You are correct that the goal is to avoid calling the destructor.  I'm not sure what you mean by "everything falling apart".  If we avoid calling destructors on these global cache objects, then we avoid everything falling apart.  It would not surprise me if there are other cases like this that need to be fixed, but we are fixing them incrementally as we find them.

The fundamental issue is that the SYCL spec does not forbid calling SYCL APIs from an application's global destructors.  Therefore, the SYCL runtime needs to allow arbitrary SYCL calls even after main() returns.  We could consider disallowing this in the SYCL spec, but I think that would be a burden on applications and it seems feasible to implement the runtime to allow it.

Reply:
> You are correct that the goal is to avoid calling the destructor. I'm not sure what you mean by "everything falling apart". If we avoid calling destructors on these global cache objects, then we avoid everything falling apart. It would not surprise me if there are other cases like this that need to be fixed, but we are fixing them incrementally as we find them.
> 
> The fundamental issue is that the SYCL spec does not forbid calling SYCL APIs from an application's global destructors. Therefore, the SYCL runtime needs to allow arbitrary SYCL calls even after main() returns. We could consider disallowing this in the SYCL spec, but I think that would be a burden on applications and it seems feasible to implement the runtime to allow it.

Yes globally this is a mess, but I am unsure about an alternative for now.
By "everything falling apart" I was thinking about what is the meaning to have a rock-solid SYCL tear-down while everything else using it could have some how exploded already... :-)
So globally keeping this as is seems good.

Reply:
BTW. I think we will need to revisit this code once https://github.com/intel/llvm/pull/2516/files is merged.